File: Homework/Objective-C/HW9/HW9/BadDoctor.m
//
//  BadDoctor.m
//  HW9
//
//  Created by Руслан on 22.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "BadDoctor.h"

@implementation BadDoctor

- (instancetype)init
{
    self = [super init];
    if (self) {
        self.headacheCount = [[NSMutableArray alloc] init];
        self.stomachacheCount = [[NSMutableArray alloc] init];
        self.soreThroatCount = [[NSMutableArray alloc] init];
    }
    return self;
}

#pragma mark - BadDoctor

- (void) clearCount {
    [_stomachacheCount removeAllObjects];
    [_headacheCount removeAllObjects];
    [_soreThroatCount removeAllObjects];
}

- (void) report {
    
    NSLog(@"          ***   BAD DOCTOR %@ REPORT ***", self.name);
    
    NSLog(@"Sore throat - %lu patient(s)", [self.soreThroatCount count]);
    NSInteger count = 0;
    for (NSString* name in _soreThroatCount) {
        count += 1;
        NSLog(@"%ld. %@", count, name);
    }
    NSLog(@"\n\n");
    count = 0;
    
    NSLog(@"Stomachache - %lu patient(s)", [self.stomachacheCount count]);
    for (NSString* name in _stomachacheCount) {
        count += 1;
        NSLog(@"%ld. %@", count, name);
    }
    NSLog(@"\n\n");
    count = 0;
    
    NSLog(@"Headache - %lu patient(s)", [self.headacheCount count]);
    for (NSString* name in _headacheCount) {
        count += 1;
        NSLog(@"%ld. %@", count, name);
    }
    NSLog(@"\n\n");
    count = 0;
}

- (void) tellTheFairyTaleFor:(Patient*)patient {
    NSLog(@"Bad doctor %@ tells the fairy tale. Patient %@ feels worse", self.name, patient.name);
}

- (void) runAway {
    NSLog(@"Bad doctor %@ runs away!", self.name);
}

- (void) request {
    NSLog(@"Can you tell the people that I make you healthy, please?))");
}

#pragma mark - PatientDelegate

- (void) patientFeelsBad:(Patient *)patient {
    NSLog(@"Patient %@ feels bad :( . (Bad doctor: %@)", patient.name, self.name);
    
    switch (patient.symptom) {
        case 0:
            [_soreThroatCount addObject:patient.name];
            NSLog(@"%@ has a sore throat", patient.name);
            NSLog(@"%@'s temperature is %f", patient.name, patient.temperature);
            
            if (patient.temperature >= 37.f && patient.temperature < 38.f) {
                [self tellTheFairyTaleFor:patient];
            } else if (patient.temperature >= 38.f && patient.temperature < 41.f) {
                [self runAway];
            }
            break;
            
        case 1:
            [_stomachacheCount addObject:patient.name];
            NSLog(@"%@ has a stomachache", patient.name);
            NSLog(@"%@'s temperature is %f", patient.name, patient.temperature);
            
            if (patient.temperature <= 38.f) {
                [self request];
                NSLog(@"Patient %@ feels worse", patient.name);
            } else if (patient.temperature > 38.f) {
                [self tellTheFairyTaleFor:patient];
            }
            break;
            
        case 2:
            [_headacheCount addObject:patient.name];
            NSLog(@"%@ has a headache", patient.name);
            NSLog(@"%@'s temperature is %f", patient.name, patient.temperature);
            
            [self tellTheFairyTaleFor:patient];
            break;
    }
}

@end

File: Homework/Swift/HW9/HW9/BadDoctor.swift
//
//  BadDoctor.swift
//  HW9
//
//  Created by Руслан on 22.09.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class BadDoctor: NSObject, PatientDelegate {
    
    let name: String
    
    init(name: String) {
        self.name = name
    }
    
// MARK: - private properties
    
    var headacheCount = [String]()
    var stomachacheCount = [String]()
    var soreThroatCount = [String]()
    
// MARK: - Bad doctor methods
    
    func tellTheFairyTale(for patient: Patient) {
        print("Bad doctor \(self.name) tells the fairy tale. Patient \(patient.name) feels worse")
    }
    
    func runAway() {
        print("Bad doctor \(self.name) runs away!")
    }
    
    func request() {
        print("Can you tell the people that I make you healthy, please?))")
    }
    
    func clearCount() {
        self.headacheCount.removeAll()
        self.stomachacheCount.removeAll()
        self.soreThroatCount.removeAll()
    }
    
    func report() {
        
        print("          *** BAD DOCTOR \(self.name.uppercased()) REPORT ***")
        
        print("Sore throat: \(soreThroatCount.count) patient(s)")
        for (index, patientName) in soreThroatCount.enumerated() {
            print("\(index + 1). \(patientName)")
        }
        print("")
        
        print("Headache: \(headacheCount.count) patient(s)")
        for (index, patientName) in headacheCount.enumerated() {
            print("\(index + 1). \(patientName)")
        }
        print("")
        
        print("Stomachache: \(stomachacheCount.count) patient(s)")
        for (index, patientName) in stomachacheCount.enumerated() {
            print("\(index + 1). \(patientName)")
        }
        print("\n\n\n")
    }

    
// MARK: - PatientDelegate
    
    func patientFeelsBad(patient: Patient) {
        
        print("Patient \(patient.name) feels bad. Bad doctor: \(self.name)")
        
        switch patient.symptom! {
            
        case .soreThroat:
            soreThroatCount.append(patient.name)
            
            print("\(patient.name) has a sore throat\n",
                "His temperature is \(patient.temperature)")
            
            switch patient.temperature {
            case 36..<38:
                self.tellTheFairyTale(for: patient)
                print("", terminator: "\n\n")
            case 38..<41:
                self.runAway()
                print("", terminator: "\n\n")
            default:
                break
            }
            
        case .stomachache:
            stomachacheCount.append(patient.name)
            
            print("\(patient.name) has a stomachache\n",
                "His temperature is \(patient.temperature)")
            
            switch patient.temperature {
            case 36..<38:
                self.request()
                print("", terminator: "\n\n")
            case 38..<41:
                self.tellTheFairyTale(for: patient)
                print("", terminator: "\n\n")
            default:
                break
            }
            
        case .headache:
            headacheCount.append(patient.name)
            
            print("\(patient.name) has a headache\n",
                "His temperature is \(patient.temperature)")
            
            self.tellTheFairyTale(for: patient)
            print("", terminator: "\n\n")
        }
    }
}
